src/vsr/journal.zig

```diff
@@ -536,20 +621,67 @@ pub fn Journal(comptime Replica: type, comptime Storage: type) type {
                 .destination_replica = destination_replica,
             };
 
-            assert(exact.offset + physical_size <= self.size_circular_buffer);
+            const buffer = message.buffer[0..size_physical];
+            const offset = offset_physical(.prepares, slot);
 
-            const buffer = message.buffer[0..physical_size];
-            const offset = self.offset_in_circular_buffer(exact.offset);
+            log.debug(
+                "{}: read_sectors: offset={} len={}",
+                .{ replica.replica, offset, buffer.len },
+            );
 
             // Memory must not be owned by `self.headers` as these may be modified concurrently:
             assert(@ptrToInt(buffer.ptr) < @ptrToInt(self.headers.ptr) or
-                @ptrToInt(buffer.ptr) > @ptrToInt(self.headers.ptr) + self.size_headers);
+                @ptrToInt(buffer.ptr) > @ptrToInt(self.headers.ptr) + headers_size);
+
+            assert_bounds(.prepares, offset, buffer.len);
+            self.storage.read_sectors(on_read, &read.completion, buffer, offset);
+        }
+
+        /// Read a prepare from disk, but there is no in-memory header.
+        pub fn read_prepare_without_header(
```

Comment:
This is great to have now.

Let's rename this to `read_prepare_with_op_and_checksum()` then add in the `// Skip the disk read if the header is all we need:` fast path from `read_prepare()` and then `read_prepare()` could call out to this after its preliminary checks.

Reply:
Then they can share quite a bit of code.

Reply:
We would just need to drop the negative assert that there's no header for the slot, but if there is a checksum then it's fine.

Reply:
They can share code, but not the `Skip the disk read if the header is all we need` path, since that requires knowing the message size — which in the `prepare_checksums` path we don't necessarily have in memory. Everything after that can be shared, though :+1:.